GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs

```diff
@@ -290,15 +262,48 @@ public void Add(GitRevision revision, int? insertScore = null, int insertRange =
 
             // Ensure all parents are loaded before adding it to the _nodes list. This is important for ordering.
             ImmutableInterlocked.Update(ref _nodes, (list, revision) => list.Add(revision), revisionGraphRevision);
+        }
 
-            if (!updateParents)
+        /// <summary>
+        /// Complete the revision info for an existing revision, not changing score or segments to parents.
+        /// This is used to add the index (staged) revision when the HEAD is not in the log.
+        /// </summary>
+        /// <param name="revision">The revision to add.</param>
+        public void AddExisting(GitRevision revision)
+        {
+            // Index has no parent segment in the graph. Add missing information.
+            if (_nodeByObjectId.TryGetValue(revision.ObjectId, out RevisionGraphRevision revisionGraphRevision))
             {
-                // The rows may already be cached, invalidate and update but do not reload rows in this context
-                _reorder = true;
-                Updated?.Invoke();
+                revisionGraphRevision.GitRevision = revision;
+                ImmutableInterlocked.Update(ref _nodes, (list, revision) => list.Add(revision), revisionGraphRevision);
             }
         }
 
+        /// <summary>
+        /// Insert a range of scores before the given score.
+        /// </summary>
+        /// <param name="insertScore">The score where to insert revisions</param>
+        /// <param name="insertRange">The range to insert</param>
+        public void ReserveRange(int insertScore, int insertRange)
+        {
+            // This will lock the cache, so it is consistent.
```

Comment:
Which "This" will lock the cache?

Reply:
Changing the comment (_nodeByObjectId.Values will give snapshot of the values but it is not relevant here)

// _nodeByObjectId is only updated from one thread so no explicit lock is needed
// _orderedNodesCache is invalidated after the revisions have been inserted